apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/MetricRegistrySerializer.java

```diff
@@ -69,29 +69,39 @@ private static void serializeSamples(Map<String, DoubleSupplier> samples, JsonWr
         final int size = samples.size();
         if (size > 0) {
             final Iterator<Map.Entry<String, DoubleSupplier>> iterator = samples.entrySet().iterator();
-            Map.Entry<String, DoubleSupplier> kv = iterator.next();
-            serializeSample(kv.getKey(), kv.getValue().get(), jw);
-            for (int i = 1; i < size; i++) {
-                jw.writeByte(JsonWriter.COMMA);
+            String key = null;
+            double value = Double.NaN;
+            Map.Entry<String, DoubleSupplier> kv;
+            while (iterator.hasNext() && !isValid(value)) {
                 kv = iterator.next();
-                serializeSample(kv.getKey(), kv.getValue().get(), jw);
+                key = kv.getKey();
+                value = kv.getValue().get();
+            }
+            if (isValid(value)) {
+                serializeSample(key, value, jw);
+                while (iterator.hasNext()) {
```

Comment:
I found the code a bit hard to follow. Especially because there is not a clear separation of where the first value is serialized and where the rest is serialized. I think a simple change would help: move the `while (iterator.hasNext())` and the corresponding scope out of the `if (isValid(value))` scope and move the serialization of the first value to the first loop. You can then also use separate local variables which are scoped inside the loop. T

```java
            // serialize first valid value
            while (iterator.hasNext() && !isValid(value)) {
                Map.Entry<String, DoubleSupplier> kv = iterator.next();
                double value = kv.getValue().get();
                if (isValid(value)) {
                    serializeSample(kv.getKey(), value, jw);
                }
            }

            // serialize rest
            while (iterator.hasNext()) {
                Map.Entry<String, DoubleSupplier> kv = iterator.next();
                double value = kv.getValue().get();
                if (isValid(value)) {
                    jw.writeByte(JsonWriter.COMMA);
                    serializeSample(kv.getKey(), value, jw);
                }
            }
```

This has equal semantics, but you can reason about the logic for serializing the first and subsequent values independently. Reducing the scope of the local variables also supports that. Another benefit is that the serialization of the first vs the subsequent values look really similar so it's easier to spot what the differences are. Another minor benefit is that this reduces the nesting level by one.